Implement clone method

veren-damyanov opened the pull request:
Resolves #10

veren-damyanov committed:
Implement clone method

ya55en committed:
Replace DepDag::vertices() access with __getitem__ and __getattr__

 Access to Depdag::_vertices is intentioanlly provided via __getitem__()
 and __getattr__(), thus the suggested rewrite which is equivalent.

 Fixed unit tests as well.

yassen-itlabs commented:
@veren-itlabs: the new `DepDag::clone()` function looks great, thank you !!

I believe the `DepDag::vertices` public exposure is not necessary and can be omitted, as the DepDag object provides access to any of its vetices via both `__getitem__` and `__getattr__`. Thus I suggest removing that and reflected this into the code.

Also, I suggest removing test checks like `self.assertIsNotNone(some_vertex)` as the payload check would blow if the vertex is `None` -- hope you are okay with this too.

veren-damyanov committed:
Implement clone cyclic dag test scenario

ya55en committed:
Add 'clone_payload_method' keyword arg to ``DepDag::clone()``

 This allows for several different usage scenarios, like:

 - do not copy the payload at all during cloning;
 - put a reference to the original payload (the default);
 - make and put a shallow copy of the original payload;
 - make and put a deep copy of the original payload.

ya55en committed:
Rename DepDag::clone() to new_vertex(); update CREDITS.rst

ya55en committed:
Fix incorrect type alias definitions

ya55en reviewed (approved):
Looks good! Ship it ;)

ya55en committed:
Merge branch 'master' into 10-clone
